base/task.jl

```diff
@@ -503,17 +503,17 @@ isolating the asynchronous code from changes to the variable's value in the curr
     Interpolating values via `\$` is available as of Julia 1.4.
 """
 macro async(expr)
-    do_async_macro(expr)
+    do_async_macro(expr, __source__)
 end
 
 # generate the code for @async, possibly wrapping the task in something before
 # pushing it to the wait queue.
-function do_async_macro(expr; wrap=identity)
+function do_async_macro(expr, linenums; wrap=identity)
     letargs = Base._lift_one_interp!(expr)
 
     thunk = esc(:(()->($expr)))
```

Comment:
IIUC, the expression you intended to alter is this async thunk, not the inline code which already executes in the correct stack frame (and may confusingly get duplicated stack frame entries with the current implementation)?
```suggestion
    thunk = Base.replace_linenums!(:(()->($esc(expr))), linenums)
```

Reply:
You're quite right, I should just be modifying the closure AST and preserving the rest